### PR TITLE
Feature unit tests

### PR DESCRIPTION
**Pull request by susrisha**
- increases test coverage to 78%
- Added unit test cases for 
     - Topic
     - FileEntity
     - Queue
     - Storage Classes
     - Logger Classes
     - Configurations
     - Authorizer
NOTE: The enumeration document is not yet complete. Here are the test results from the above code:
```shell
> nodets-ms-core@0.0.8 test
> jest --silent

 PASS  test/local_logger.unit.ts
  Local logger
    ✓ Should initialize without parameters (1 ms)
    ✓ Should send add request (2 ms)
    ✓ Should be able to  update request
    ✓ Should be able to  add event
    ✓ Should be able to record metric
    ✓ Should record HTTP call
    ✓ Should send info (1 ms)
    ✓ Should send debug
    ✓ Should record message
    ✓ Should get Auditor

 PASS  test/queue.unit.ts
  Queue 
    ✓ Should get the Azure Queue (5 ms)
    ✓ Should be able to add messages (1 ms)
    ✓ Should be able to send messages
    ✓ Should call listen messages

 PASS  test/azure_storage_client.unit.ts
  Azure Storage Client
    ✓ Should initialize Client (1 ms)
    ✓ Should get Container with name
    ✓ Should get file from the container name and entity (1 ms)
    ✓ Should get file from URL

 PASS  test/topic.unit.ts
  Azure service bus topic
    ✓ Should have provider as azure (1 ms)
    ✓ Should get the service topic (1 ms)
    ✓ Should send message via ServiceBus when message is sent
    ✓ Should listen to the said subscription appropriately
    ✓ Should throw error when the client is not available (10 ms)
  Local topic
    ✓ Should have provider as Local (1 ms)
    ✓ Should get the service topic
    ✓ Should send message via ServiceBus when message is sent
    ✓ Should listen to the said subscription appropriately
    ✓ Should throw error when the client is not available (2 ms)

 PASS  test/azure_auditor.unit.ts
  Azure auditor
    ✓ Should initialize with queue name (1 ms)
    ✓ Should add event appropriately (1 ms)
    ✓ Should add request  (1 ms)
    ✓ Should update request 
    ✓ Should return object with empty queue also

 PASS  test/azure_storage_container.unit.ts
  Azure Storage container
    ✓ Should be able to initialize from ContainerClient (1 ms)
    ✓ Should be able to list files based on response from azure (1 ms)
    ✓ Should be able to create a file with given name (1 ms)

 PASS  test/azure_analytic.unit.ts
  Azure analytic
    ✓ Should initialize with queue name (1 ms)
    ✓ Should be able to send message (2 ms)

 PASS  test/local-topic.unit.ts
  Local topic unit test
    ✓ Should initialize with any config (12 ms)
    ✓ Should initialize without any config (1 ms)
    ✓ Should publish message to channel with publish (319 ms)
    ✓ Should consume from the channel with consume (303 ms)

 PASS  test/core.unit.ts
  Core unit tests
    ✓ Should initialize with default configuration (1 ms)
    ✓ Should initialize with Local Configuration (1 ms)
    ✓ Should get default storage client by default (8 ms)
    ✓ Should get Local storage client for local configuration
    ✓ Should get Local storage client for local Configuration
    ✓ Should get Azure storage client for azure config
    ✓ Should return Local storage client if core initialized by local and default storage requested (1 ms)
    ✓ Should get Azure logger by default
    ✓ Should return local logger for local
    ✓ Should return Hosted Authorizer if provider not given (3 ms)

 PASS  test/queue-message.unit.ts
  QueueMessage
    ✓ Should create Queue appropriately (5 ms)
    ✓ Should Instantiate

 PASS  test/logger.unit.ts
  Logger
    ✓ Should initialize with default configuration (11 ms)
    ✓ Should send info (1 ms)
    ✓ Should send debug
    ✓ Should send metric (1 ms)
    ✓ Should record HTTP call (8 ms)
    ✓ Should get auditor (5 ms)
    ✓ Should get analytic (4 ms)
    ✓ Should call send when sendAll called

 PASS  test/local_file_entity.unit.ts
  Local file entity
    ✓ Should get the file appropriately (1 ms)
    ✓ Should get the file from axios (30 ms)
    ✓ Should get the data text (2 ms)
    ✓ Should upload from data (1 ms)

 PASS  test/azure_file_entity.unit.ts
  Azure file entity
    ✓ Should be able to initialize
    ✓ Should be able to download when the file stream is needed
    ✓ Should be able to get the text (6 ms)
    ✓ Should be able to upload the content from readable stream (1 ms)

 PASS  test/azure-service-bus-topic.unit.ts
  Azure service bus topic unit
    ✓ Should initialize with Azure default config if not provided (1 ms)
    ✓ Should have called and created topic and sender on init
    ✓ Should listen to a subscription using subscribe method (1 ms)
    ✓ Should be able to call publish on sender when message is sent

 PASS  test/azure-service-bus-queue.unit.ts
  Azure service bus
    ✓ should intialize with proper queue (2 ms)
    ✓ Should be able to send messages (6 ms)
    ✓ Should be able to receive messages

 PASS  test/local_storage_container.unit.ts
  Local storage container
    ✓ Should be able to create with server root and name (1 ms)
    ✓ Should get list of files as given by server (1 ms)
    ✓ Should create a new file based on path

 PASS  test/core_config.unit.ts
  Core Configuration
    ✓ Should initialize Core default with Azure (2 ms)
    ✓ Should initialize Core default based on env variable

 PASS  test/local_storage_client.unit.ts
  Local storage client
    ✓ Should initialize Client appropriately (1 ms)
    ✓ Should return the container
    ✓ Should get file from container (1 ms)
    ✓ Should get the file from the URL
    ✓ Should throw a Not found error when not found (3 ms)

 PASS  test/azure_logger_config.unit.ts
  Azure logger Config
    ✓ Should have provider as azure (1 ms)
    ✓ Should pickup queuename from local configuration
    ✓ Should pickup queuename from parameter

 PASS  test/local_storage_config.unit.ts
  Local Storage config
    ✓ Should initialize with no configuration given
    ✓ Should take the env connection
    ✓ Should take values from default when default is called

 PASS  test/azure_storage_config.unit.ts
  Azure storage config
    ✓ Should initialize with given parameters (1 ms)
    ✓ Should get the value of connection string from process
    ✓ Default should pick up from env (1 ms)

 PASS  test/authorizer.int.ts (7.723 s)
  Authorizer (Hosted and Simulated)
    ✓ Should give HostedAuthorizer for configuration (2 ms)
    ✓ Should give SimulatedAuthorizer for configuration
    ✓ Should pick up host url from env if not given
    ✓ Should configure based on the url provided
    ✓ Should give error for no permissions (2 ms)
    ✓ Should give simulated result as false when userId is as per predefined
    ✓ Should give simulated result as expected (true)
    ✓ Should respond true when user has permission (1695 ms)
    ✓ Should respond false if user does not have permission (1237 ms)
    ✓ Should respond true if user has all the permissions required (1127 ms)
    ✓ Should reject if the URL is malformed (34 ms)

----------------------------------|---------|----------|---------|---------|-----------------------------------------------------------
File                              | % Stmts | % Branch | % Funcs | % Lines | Uncovered Line #s                                         
----------------------------------|---------|----------|---------|---------|-----------------------------------------------------------
All files                         |   78.18 |    61.42 |   78.39 |   78.05 |                                                           
 src                              |   77.77 |    68.18 |     100 |   77.77 |                                                           
  core.ts                         |   77.77 |    68.18 |     100 |   77.77 | 30,41,60,76,96-97,105-111,121-122,152,163-168,186,197,204 
 src/core/auth/model              |     100 |      100 |     100 |     100 |                                                           
  auth_config.ts                  |     100 |      100 |     100 |     100 |                                                           
  permission_request.ts           |     100 |      100 |     100 |     100 |                                                           
 src/core/auth/provider/hosted    |     100 |      100 |     100 |     100 |                                                           
  hosted_authorizer.ts            |     100 |      100 |     100 |     100 |                                                           
 src/core/auth/provider/simulated |     100 |      100 |     100 |     100 |                                                           
  simulated_authorizer.ts         |     100 |      100 |     100 |     100 |                                                           
 src/core/logger                  |     100 |    76.19 |    90.9 |     100 |                                                           
  index.ts                        |     100 |      100 |     100 |     100 |                                                           
  logger.ts                       |     100 |    76.19 |      90 |     100 | 23-28,57-82                                               
 src/core/logger/abstracts        |     100 |      100 |     100 |     100 |                                                           
  logger-abstract.ts              |     100 |      100 |     100 |     100 |                                                           
 src/core/logger/model            |     100 |      100 |     100 |     100 |                                                           
  audit_event.ts                  |     100 |      100 |     100 |     100 |                                                           
  audit_request.ts                |     100 |      100 |     100 |     100 |                                                           
 src/core/logger/providers        |     100 |    81.25 |     100 |     100 |                                                           
  azure_analytic.ts               |     100 |      100 |     100 |     100 |                                                           
  azure_auditor.ts                |     100 |    76.92 |     100 |     100 | 18-41                                                     
  azure_logger_config.ts          |     100 |      100 |     100 |     100 |                                                           
 src/core/logger/providers/local  |   85.18 |      100 |      80 |   85.18 |                                                           
  local_logger.ts                 |   85.18 |      100 |      80 |   85.18 | 59,82,96-97                                               
 src/core/queue                   |   77.77 |    58.82 |      80 |   76.47 |                                                           
  index.ts                        |     100 |      100 |     100 |     100 |                                                           
  queue.ts                        |   62.96 |    36.36 |    62.5 |   62.96 | 24,32,40-41,50,55-67                                      
  topic.ts                        |   90.47 |      100 |     100 |   90.47 | 41,48                                                     
 src/core/queue/abstracts         |     100 |      100 |     100 |     100 |                                                           
  message-queue.ts                |     100 |      100 |     100 |     100 |                                                           
  message-topic.ts                |     100 |      100 |     100 |     100 |                                                           
 src/core/queue/decorators        |      25 |        0 |       0 |      25 |                                                           
  when.decorator.ts               |      25 |        0 |       0 |      25 | 8-13                                                      
 src/core/queue/models            |     100 |      100 |     100 |     100 |                                                           
  queue-message.ts                |     100 |      100 |     100 |     100 |                                                           
 src/core/queue/providers         |   76.81 |       50 |   88.23 |   76.47 |                                                           
  azure-queue-config.ts           |     100 |      100 |     100 |     100 |                                                           
  azure-service-bus-queue.ts      |   79.06 |    22.22 |     100 |   78.57 | 43,68,87-96                                               
  azure-service-bus-topic.ts      |   68.18 |       80 |      60 |   68.18 | 33-40,46                                                  
 src/core/queue/providers/local   |   40.57 |    44.89 |   38.88 |   41.17 |                                                           
  local-queue-config.ts           |     100 |      100 |     100 |     100 |                                                           
  local-queue.ts                  |      10 |        0 |       0 |   10.25 | 18-94                                                     
  local-topic.ts                  |      80 |    74.07 |   83.33 |      80 | 42-48                                                     
 src/core/storage/providers/azure |   88.88 |    66.66 |   95.23 |   89.85 |                                                           
  azure_file_entity.ts            |   91.66 |      100 |     100 |   95.45 | 70                                                        
  azure_storage_client.ts         |   78.57 |        0 |    87.5 |   78.57 | 48-59                                                     
  azure_storage_config.ts         |     100 |      100 |     100 |     100 |                                                           
  azure_storage_container.ts      |     100 |      100 |     100 |     100 |                                                           
 src/core/storage/providers/local |   97.22 |      100 |   91.66 |   97.14 |                                                           
  local_file_entity.ts            |     100 |      100 |     100 |     100 |                                                           
  local_storage_client.ts         |   95.83 |      100 |   88.88 |   95.83 | 37                                                        
  local_storage_config.ts         |     100 |      100 |     100 |     100 |                                                           
  local_storage_container.ts      |   94.11 |      100 |    87.5 |   93.75 | 36                                                        
 src/decorators                   |   59.09 |    31.25 |      25 |      55 |                                                           
  nested-model.decorator.ts       |   30.76 |        0 |       0 |      25 | 17-33,45-48                                               
  prop.decorator.ts               |     100 |    83.33 |     100 |     100 | 33                                                        
 src/models                       |     100 |      100 |      80 |     100 |                                                           
  config.ts                       |     100 |      100 |     100 |     100 |                                                           
  index.ts                        |     100 |      100 |   66.66 |     100 |                                                           
 src/models/base                  |   53.42 |     61.9 |   56.25 |   54.16 |                                                           
  abstract-domain-entity.ts       |   53.42 |     61.9 |   56.25 |   54.16 | 56,71-85,137,140,157-226                                  
 src/utils/resource-errors        |      84 |       70 |      75 |      84 |                                                           
  abstract-resource-error.ts      |   81.25 |       70 |     100 |   81.25 | 47,52,56                                                  
  bad-request-resource-error.ts   |     100 |      100 |     100 |     100 |                                                           
  not-found-resource-error.ts     |     100 |      100 |     100 |     100 |                                                           
  unprocessable-resource-error.ts |   66.66 |      100 |       0 |   66.66 | 5                                                         
 test                             |   84.21 |      100 |   72.72 |   84.21 |                                                           
  azure.mock.ts                   |     100 |      100 |     100 |     100 |                                                           
  rabbitmq.mock.ts                |   66.66 |      100 |   57.14 |   66.66 | 7,72-78                                                   
----------------------------------|---------|----------|---------|---------|-----------------------------------------------------------
Test Suites: 22 passed, 22 total
Tests:       107 passed, 107 total
Snapshots:   0 total
Time:        8.154 s, estimated 12 s

```

### CODE REVIEW

File: src/core.ts [hunk 1]
```diff
@@ -109,6 +110,12 @@ export class Core {
                     console.debug(config);
                     return null;
                 }
+            } else if(config.provider == 'Local'){
```
Comment (MashB) [inline src/core.ts]: 'Local' , if we can use CONST variable will help

File: src/core/queue/providers/azure-service-bus-queue.ts [hunk 2]
```diff
@@ -56,13 +56,17 @@ export class AzureServiceBusQueue implements IMessageQueue {
         this.currentMessages.push(message);
     }
 
-    private async listenQueue() {
+    private async listenQueue(always:boolean = true) {
         this.listener.receiveMessages(1).then((messages) => {
             messages.forEach(async (singleMessage) => {
+                console.log('Received message');
```
Comment (MashB) [inline src/core/queue/providers/azure-service-bus-queue.ts]: Remove console logs. sensitive messages will have concerns

File: test-report.html
Comment (MashB) [inline test-report.html:1]: this file can be git ignored

File: test/authorizer.int.ts [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+import { HostedAuthorizer } from "../src/core/auth/provider/hosted/hosted_authorizer";
+import { SimulatedAuthorizer } from "../src/core/auth/provider/simulated/simulated_authorizer";
+
+require('dotenv').config()
+
+describe('Authorizer (Hosted and Simulated)', () => {
+
+    it('Should give HostedAuthorizer for configuration', () => {
+        // Arrange and act
+        const authorizer = Core.getAuthorizer({ provider: "Hosted" });
```
Comment (MashB) [inline test/authorizer.int.ts]: if we can use const enum for Provider will help. Applies to all such occurrences

File: test/authorizer.int.ts [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+            userId: "abc",
+            orgId: "sdfs",
+            permssions: [],
+            shouldSatisfyAll: true
+        });
+        // Act and assert
+        await expect(authorizer?.hasPermission(permission)).resolves.toEqual(true);
+    });
+
+    it('Should respond true when user has permission', async () => {
```
Comment (MashB) [inline test/authorizer.int.ts]: for all Integration test create section (describe) or create file .integration.ts and move all tests

**Review by MashB — CHANGES_REQUESTED** [review]

**Review by MashB — APPROVED** [review]

**Review by sujata-m — APPROVED** [review]

**Pull request merged**